pkg/compiler/validators/bindings.go

```diff
@@ -125,6 +125,26 @@ func validateBinding(w c.WorkflowBuilder, nodeID c.NodeID, nodeParam string, bin
 					}
 				}
 
+				// If the type is a struct (e.g. dataclass) and the attribute path is longer than 0,
+				// We skip the type check and let it fail at runtime because we don't know the type of struct field
+				if sourceType.GetSimple() == flyte.SimpleType_STRUCT && len(val.Promise.AttrPath) > 0 {
```

Comment:
@wild-endeavor is this necessarily true. if we have `dataclass_json` do we know the type?

Reply:
@ByronHsu would you mind just trying to fill this in a bit?  If it's really not doable, or if it has a bunch of edge cases, I think it's okay.

Walking through some examples.  Let's say the simple case of
```python
@task
def t1() -> List[str]:
    ...

@task
def t2(needs_str: str):
    ...

@workflow
def wf():
    res = t1()
    t2(needs_str=res[5])
```
In this case `t1` will have a TypedInterface output of
```
        value {
          type {
            collection_type {
              simple: STRING
            }
          }
          description: "o0"
        }
```

and because it's a string it will type check against the string in `t2`.  The same can be applied for nested lists and dictionaries.  As long as they are expressed via `collection_type` and `map_value_type` they are relatively easily type-checkable. I assume this is what the code below is doing?


The dataclass case is more complicated
```python
@dataclass_json
@dataclass
class MyDC(object):
    snapshotDate: datetime
    region: str

@task
def t1(needs_dt: datetime):
    ...

@task
def t2(needs_str: str):
    ...

@workflow
def wf(a: MyDC):
    t1(needs_dt=a.snapshotDate)
    t2(needs_str=a.region)
```

The reason it's more complicated is because the dataclass types are completely obscured (esp. since flyte idl currently doesn't support multi-variate map types).  I assume this is why you're skipping checking in the simple/struct case.

Could you see if it's possible though to capture it?  Can we
* Add a new field to LiteralType that is only relevant for the scalar case.  Maybe just in `LiteralType` or in `TypeStructure`.
* Add a nested literal map of the types that's only present in the dataclass case.  Just iterate through the fields in dataclass and recursively call the TypeEngine.
* Add the same checking logic in propeller as the normal `map_value_type` if it's a simple_struct and this new field is present.

What do you think?  It will add to the correctness of this new feature.  And it will make Dan happy.  And in the end, that's what we're all really about.